### PR TITLE
FEAT-#1844: upgrade pyarrow to 1.0

### LINKED ISSUE #1844: Upgrade to Pyarrow 1.0
In #1377 we require `pyarrow<0.17`, but since Pyarrow 1.0 released, we should update to new stable API.

### PR DESCRIPTION
**Pull request by anmyachev**
Signed-off-by: Anatoly Myachev <anatoly.myachev@intel.com>

## What do these changes do?

- [ ] commit message follows format outlined [here](https://modin.readthedocs.io/en/latest/developer/contributing.html)
- [ ] passes `flake8 modin`
- [ ] passes `black --check modin`
- [ ] signed commit with `git commit -s` 
- [x] Resolves #1844  
- [ ] tests added and passing

### CODE REVIEW

**Comment by anmyachev** [comment]
```python
modin/pandas/test/test_io.py *** Aborted at 1604068767 (unix time) try "date -d @1604068767" if you are using GNU date ***
PC: @                0x0 (unknown)
*** SIGSEGV (@0x0) received by PID 10045 (TID 0x7fafa1a84700) from PID 0; stack trace: ***
    @     0x7fb061a4b890 (unknown)
    @     0x7fafedf20420 arrow::py::ConvertNumericNullableCast<>()
    @     0x7fafedf21b63 arrow::py::FloatWriter<>::CopyInto()
    @     0x7fafedf049a1 arrow::py::PandasWriter::Write()
    @     0x7fafedf07318 _ZZN5arrow2py24ConsolidatedBlockCreator18WriteTableToBlocksEvENKUliE_clEi
    @     0x7fafedf0771d _ZNSt17_Function_handlerIFvvEZN5arrow8internal8Executor6SubmitIRZNS1_2py24ConsolidatedBlockCreator18WriteTableToBlocksEvEUliE_JRiENS1_6StatusENS2_6detail20ExecutorResultTraitsISA_EESA_EENS1_6ResultINS1_6FutureIT3_EEEENS2_9TaskHintsEOT_DpOT0_E4TaskE9_M_invokeERKSt9_Any_data
    @     0x7fafdd2a0777 _ZNSt6thread11_State_implINS_8_InvokerISt5tupleIJZN5arrow8internal10ThreadPool21LaunchWorkersUnlockedEiEUlvE_EEEEE6_M_runEv
    @     0x7facac59d067 execute_native_thread_routine
    @     0x7fb061a406db start_thread
    @     0x7fb06176988f clone
Segmentation fault (core dumped)

```

File: modin/experimental/engines/pyarrow_on_ray/frame/partition.py
Comment (vnlitvinov) [inline modin/experimental/engines/pyarrow_on_ray/frame/partition.py:37]: Could you please explain the logic? I don't like workarounds, especially those which hide issues which I don't understand a root cause for.

**Review by vnlitvinov — COMMENTED** [review]
Please also either turn off omnisci tests or mark them as `xfail` until we publish omnisci with pyarrow 1.0 as dependency.

**Comment by Garra1980** [comment]
I'd say 

> Please also either turn off omnisci tests or mark them as `xfail` until we publish omnisci with pyarrow 1.0 as dependency.

I'd say we will merge this PR right after we publish new omnisci

File: modin/experimental/engines/pyarrow_on_ray/frame/partition.py
Comment (Garra1980) [reply modin/experimental/engines/pyarrow_on_ray/frame/partition.py:37]: > Could you please explain the logic? I don't like workarounds, especially those which hide issues which I don't understand a root cause for.

This was done to make sure we don't have other problems. As for given changes - for some reasons there is a segfault only on Linux for specific data conversion during arrow_to_pandas conversion. The reason is not clear at the moment

**Comment by anmyachev** [comment]
Reproducer for segmentation fault:
```python
import pandas
import pyarrow
import ray
ray.init()
 
from modin.pandas.test.test_io import _make_csv_file
 
test_file = "pyarrow_test.csv"
filenames_list = []
rows_number = 32
_make_csv_file(filenames_list)(test_file, row_size=rows_number)


df = pandas.read_csv(test_file)

oid = ray.put(df)
df_from_ray_store = ray.get(oid)

table = pyarrow.Table.from_pandas(df)
table_from_ray_store = pyarrow.Table.from_pandas(df_from_ray_store)

table.equals(table_from_ray_store, check_metadata=True)
print("Tables are equal\n", flush=True)


print(f"without ray store: {table.to_pandas().dtypes}\n", flush=True)

# failed with segmentation fault
print(f"with usage of ray store: {table_from_ray_store.to_pandas().dtypes}", flush=True)

```

**Comment by Garra1980** [comment]
rows_number = 24 - everything works...

**Comment by Garra1980** [comment]
Seems the issue is only for double so one column of floating point numbers is enough to reproduce

**Comment by Garra1980** [comment]
Unaligned access for some reasons:
=> 0x00007fffe6351420 <+736>:   movapd (%r11,%rax,1),%xmm0
   0x00007fffe6351426 <+742>:   add    $0x1,%rdx
(gdb) p /x $r11+$rax
$7 = 0x7ffb65100354
(gdb)

**Comment by anmyachev** [comment]
> Unaligned access for some reasons:
> => 0x00007fffe6351420 <+736>: movapd (%r11,%rax,1),%xmm0
> 0x00007fffe6351426 <+742>: add $0x1,%rdx
> (gdb) p /x $r11+$rax
> $7 = 0x7ffb65100354
> (gdb)

What about of using intel style? This is pretty hard to read.
Just make: `set disassembly-flavor intel`.

File: modin/experimental/engines/pyarrow_on_ray/frame/partition.py
Comment (vnlitvinov) [reply modin/experimental/engines/pyarrow_on_ray/frame/partition.py:37]: According to disassembly you posted below I would say we're just getting lucky on Windows that nothing crashes. I would prefer us to investigate the issue instead of sweeping it under the rug by using this workaround (which is wasting quite a few CPU time along the way).

**Comment by ienkovich** [comment]
I tried to build `libarrow_python.so` locally from `apache-arrow-1.0.0` tag and copy it to conda env. This resolved the issue. Conda version was built by quite old GCC version 4.4.7. A compiler bug is a possible reason for this failure. We have `double` data aligned by 4 which is legal according to ABI.

**Comment by ienkovich** [comment]
Looks like I was wrong about alignment requirements. On x86_64 double should be 8-byte aligned. So we might say the arrow is fed with the wrong input.

**Comment by ienkovich** [comment]
With more disasm analysis I found that code is perfectly fine but requires 8-byte alignment for input data. We provide 4-byte aligned data that violates x86-64 ABI.

File: modin/experimental/engines/pyarrow_on_ray/frame/partition.py
Comment (Garra1980) [reply modin/experimental/engines/pyarrow_on_ray/frame/partition.py:37]: > According to disassembly you posted below I would say we're just getting lucky on Windows that nothing crashes. I would prefer us to investigate the issue instead of sweeping it under the rug by using this workaround (which is wasting quite a few CPU time along the way).

In progress...

**Comment by ienkovich** [comment]
Here is what I got from the debugger:
```
#19 0x00007fffdc3f446a in arrow::py::NdarrayToArrow (pool=0x7ff69a5ea6f0 <arrow::jemalloc_pool>, ao=0x7ff66ca4b8a0,
    mo=0x5555558c4360 <_Py_NoneStruct>, from_pandas=true, type=..., cast_options=..., out=0x7fffffffb9d0)
    at /localdisk/ilyaenko/arrow/cpp/src/arrow/python/numpy_to_arrow.cc:839
839       RETURN_NOT_OK(converter.Convert());
(gdb) x/2fg PyArray_DATA((PyArrayObject*)(ao))
0x7ff6b36a078c: 8360.2914072769854      6172.7386846336194
```

The data matches input CSV file:
```
col1,col2,col3,col4,col5,col6
0,2000-01-01,0,00:00:00,DJSfUncxsL,8360.291407276985
10,2000-01-01,10,01:00:00,HiDFmrRvdw,6172.738684633619
...
```

That means we get an unaligned `double` array from python and it happens only for deserialized pandas frame.

**Comment by ienkovich** [comment]
Here is a problem reproducer which doesn't use pyarrow:
```
import pandas
import numpy as np
import ray
ray.init()

from modin.pandas.test.test_io import _make_csv_file

test_file = "1.csv"
filenames_list = []
rows_number = 32
_make_csv_file(filenames_list)(test_file, row_size=rows_number)

df = pandas.read_csv(test_file)

oid = ray.put(df)
df_from_ray_store = ray.get(oid)

ptr1 = df["col6"].to_numpy().__array_interface__["data"][0]
print(f"Original address: 0x{hex(ptr1)} - {'un' if ptr1 % 8 != 0 else ''}aligned")
assert(ptr1 % 8 == 0)
ptr2 = df_from_ray_store["col6"].to_numpy().__array_interface__["data"][0]
print(f"Deserialized address: 0x{hex(ptr2)} - {'un' if ptr2 % 8 != 0 else ''}aligned")
assert(ptr2 % 8 == 0)
```

**Comment by ienkovich** [comment]
Reproducer without Pandas:
```
import numpy as np
import ray
ray.init()

arr1 = np.array([10.1, 20.2, 30.3], dtype = np.double)
oid = ray.put(arr1)
arr2 = ray.get(oid)

ptr1 = arr1.__array_interface__["data"][0]
print(f"Original address: 0x{hex(ptr1)} - {'un' if ptr1 % 8 != 0 else ''}aligned")
assert(ptr1 % 8 == 0)
ptr2 = arr2.__array_interface__["data"][0]
print(f"Deserialized address: 0x{hex(ptr2)} - {'un' if ptr2 % 8 != 0 else ''}aligned")
assert(ptr2 % 8 == 0)
```

**Comment by ienkovich** [comment]
Filled ray-project/ray#11760

File: modin/pandas/test/test_io.py
Comment (vnlitvinov) [inline modin/pandas/test/test_io.py:509]: we should skip this test altogether, because marking it as `xfail` does not prevent it from running and crashing the whole testing run, which renders CI red anyway

**Review by vnlitvinov — CHANGES_REQUESTED** [review]

File: modin/pandas/test/test_io.py
Comment (anmyachev) [reply modin/pandas/test/test_io.py]: Ok, changed.

**Review by vnlitvinov — APPROVED** [review]
LGTM if CI passes (with omnisci tests expected to fail currently due to no omnisci packages with `arrow=1.0` available)

**Comment by Garra1980** [comment]
Let's wait green CI here before merging

**Review by devin-petersohn — APPROVED** [review]

**Pull request merged**
